docs/design/proposals/encryption-with-vault-tokens.md

```diff
@@ -0,0 +1,109 @@
+# Multi-tenancy with Vault Tokens
+
+## Current Feature: Vault access with single token
+
+The current feature to [support Hashicorp Vault as a KMS is
+documented](./encrypted-pvc.md):
+
+- tenants can create PVCs that are encrypted with a unique key
+- the key to encrypt/decrypt the PVC is stored in Hashicorp Vault
+- there is a single token (Kubernetes ServiceAccount) to access Vault
+- enabling is done in the StorageClass, pointing to a KMS configuration section
+  in a JSON configuration file (Kubernetes ConfigMap)
+
+## High-Level Requirements
+
+The new feature should support multi-tenancy, so that each Tenant can use their
+own Vault Token for PVC encryption:
+
+- each tenant can manage their Vault token (and PVC keys)
+- to get the key from Vault, each tenant should provide their personal token
+- Ceph-CSI should use the token from the tenant to store the unique key for the
+  PVC
+- a Tenant is the owner of a Kubernetes Namespace
+
+## Restrictions to consider
+
+- tenants can only configure their Kubernetes namespace, their token to access
+  Vault should be located in their namespace (as a Kubernetes Secret)
+- Ceph-CSI can not directly access Kubernetes namespaces, as CSI is an
+  abstraction layer
+- Ceph-CSI needs to talk to a service (sidecar) to access the token from a
+  tenant
+- the KMS configuration is a ConfigMap with name
+  `ceph-csi-encryption-kms-config` in the namespace where the Ceph-CSI pods are
+  running
+- the KMS configuration is available for the Ceph-CSI pods at
+  `/etc/ceph-csi-encryption-kms-config/config.json`
+- [example of the
+  configuration](https://github.com/ceph/ceph-csi/blob/master/examples/kms/vault/kms-config.yaml)
+
+## Dependencies
+
+- the name of the Kubernetes Secret needs to be known (configured in the
+  StorageClass, or somewhere else)
+- each tenant (Kubernetes namespace) should use the same name for the Secret
+  that contains the Vault token
+- the Ceph-CSI sidecar needs to store the namespace (and name of the Secret?)
+  together with the metadata of the PVC
+
+## Implementation Outline
+
+- the Ceph-CSI node-plugin needs to request the token in the NodeStage CSI
+  operation and create/get the key for the PVC
+- the Ceph-CSI provisioner needs to request the token to delete the key for the
+  PVC in the VolumeDelete CSI operation
+- when creating the PVC the Ceph-CSI provisioner needs to store the namespace
+  (and Kubernetes Secret?) of the PVC in its metadata
+  - stores the `csi.volume.owner` (name of Tenant) in the journal and sets it
```

Comment:
Sure, the location does not really matter much to me. There also is some metadata for encryption added to the image, so adding the owner makes sense. Any suggestion for other encyption details? (type, or ID of the KMS maybe?)

#1093 sets it in the journal at the moment, but that is mostly for the e2e testing.

Reply:
okay we need to consider both cephfs and rbd both as we may support encryption for cephfs also, keeping it at omap is fine, the omap generator is already generating the required omap, we need to update the controller to add owner also.

Reply:
we are already setting KMS ID on omap data in ReserveName. what other informations do we need to store?

Reply:
I can not think of anything else either. Owner and KMS ID are the ones that should be sufficient.

Reply:
In case of RBD mirroring, when we get the static PV created it should have the claimRef which can be used to re-populate the OMap regarding the namespace to fetch the token from?

Reply:
Yes, we can make use of a namespace in claimRef which is part of PV spec.